### PR TITLE
feat(api): add support for `table_expr.nunique(where=...)`

### LINKED ISSUE #6683: feat: support nunique on tables
### Is your feature request related to a problem?

currently, `.nunique()` is only defined on columns, so I have to do `.distinct().count()`, which is way more typing. it could be nice if `.nunique()` were supported on tables, but there could be performance/other implications

### Describe the solution you'd like

`t.select("some_column").nunique()`

### What version of ibis are you running?

master

### What backend(s) are you using, if any?

duckdb but any

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Add support for `table_expr.nunique(where=...)`. Closes #6683.

### CODE REVIEW

File: ibis/backends/tests/test_aggregation.py [hunk 1]
```diff
@@ -772,6 +772,64 @@ def test_reduction_ops(
...
+@pytest.mark.parametrize(
+    ("ibis_cond", "pandas_cond"),
+    [
+        param(lambda _: None, lambda _: slice(None), id="no_cond"),
+        param(
+            lambda t: t.string_col.isin(["1", "7"]),
+            lambda t: t.string_col.isin(["1", "7"]),
+            id="cond",
+            marks=[
+                mark.broken(["datafusion"], raises=AssertionError),
```
Comment (gforsyth) [inline ibis/backends/tests/test_aggregation.py:784]: This seems redundant given the `notyet` marker for `datafusion` below on line 820

File: ibis/backends/tests/test_aggregation.py [hunk 2]
```diff
@@ -772,6 +772,64 @@ def test_reduction_ops(
...
+@mark.notyet(
+    ["mssql"],
+    raises=(sa.exc.ProgrammingError, com.UnsupportedOperationError),
+    reason="backend doesn't support count distinct with multiple columns",
+)
+@mark.notyet(
+    ["oracle"],
+    raises=(sa.exc.DatabaseError, com.UnsupportedOperationError),
+    reason="oracle doesn't support count distinct with multiple columns",
+)
```
Comment (gforsyth) [inline ibis/backends/tests/test_aggregation.py:818]:
```suggestion
@mark.notyet(
    ["bigquery", "druid", "mssql", "oracle", "sqlite"],
    raises=(sa.exc.OperationalError, sa.exc.DatabaseError, com.UnsupportedOperationError),
    reason="backend doesn't support count distinct with multiple columns",
)
```
??

File: ibis/expr/types/relations.py [hunk 3]
```diff
@@ -2012,6 +2012,21 @@ def filter(
...
+        Parameters
+        ----------
+        where
+            Optional boolean expression to filter rows when counting.
+
+        Returns
+        -------
+        IntegerScalar
+            Number of unique rows in the table
+        """
```
Comment (jcrist) [inline ibis/expr/types/relations.py:2027]: Could you add a small example here as well (since all our other `Table` methods have examples)?

**Review by gforsyth — APPROVED** [review]
This looks good!  Two small nits in the test layout, but nothing blocking.

**Comment by kszucs** [comment]
I think we should model `CountStar` and `CountDistinctStar` differently since they don't operate on single values but rather on relations.

File: ibis/backends/polars/compiler.py [hunk 4]
```diff
@@ -1157,3 +1157,11 @@ def execute_view(op, *, ctx: pl.SQLContext, **kw):
 @translate.register(ops.SelfReference)
 def execute_self_reference(op, **kw):
     return translate(op.table, **kw)
+
+
+@translate.register(ops.CountDistinctStar)
+def execute_count_distinct_star(op, **kw):
+    arg = pl.struct(*op.arg.schema.names)
+    if (where := op.where) is not None:
+        arg = arg.filter(translate(where, **kw))
```
Comment (jcrist) [inline ibis/backends/polars/compiler.py:1166]: IMO the use of the `:=` doesn't add a benefit here (attribute access is cheap, no reason to only evaluate it once) and makes the code harder to read. Definitely opinion based though, up to you.

```suggestion
    if op.where is not None:
        arg = arg.filter(translate(op.where, **kw))
```

File: ibis/backends/clickhouse/compiler/values.py [hunk 5]
```diff
@@ -1412,3 +1412,12 @@ def _array_zip(op: ops.ArrayZip, **kw: Any) -> str:
             sql_arg = sql_arg.sql(dialect="clickhouse")
         arglist.append(sql_arg)
     return f"arrayZip({', '.join(arglist)})"
+
+
+@translate_val.register(ops.CountDistinctStar)
+def _count_distinct_star(op: ops.CountDistinctStar, **kw: Any) -> str:
+    column_list = ", ".join(map(_sql, map(sg.column, op.arg.schema.names)))
+    if (where := op.where) is not None:
+        return f"countDistinctIf(({column_list}), {translate_val(where, **kw)})"
```
Comment (jcrist) [inline ibis/backends/clickhouse/compiler/values.py:1421]:
```suggestion
    if op.where is not None:
        return f"countDistinctIf(({column_list}), {translate_val(op.where, **kw)})"
```

File: ibis/backends/base/sql/alchemy/registry.py [hunk 6]
```diff
@@ -493,13 +494,50 @@ def _count_star(t, op):
...
+    if t._has_reduction_filter_syntax:
+        return func(*cols).filter(t.translate(where))
+
+    if not t._supports_tuple_syntax and len(cols) > 1:
+        raise com.UnsupportedOperationError(
+            f"{t._dialect_name} backend doesn't support `COUNT(DISTINCT ...)` with a "
+            "filter with more than one column"
+        )
+
+    return sa.func.count(t.translate(ops.Where(where, sa.distinct(*cols), None)))
```
Comment (jcrist) [inline ibis/backends/base/sql/alchemy/registry.py:518]:
```suggestion
    if op.where is None:
        return func(*cols)

    if t._has_reduction_filter_syntax:
        return func(*cols).filter(t.translate(op.where))

    if not t._supports_tuple_syntax and len(cols) > 1:
        raise com.UnsupportedOperationError(
            f"{t._dialect_name} backend doesn't support `COUNT(DISTINCT ...)` with a "
            "filter with more than one column"
        )

    return sa.func.count(t.translate(ops.Where(op.where, sa.distinct(*cols), None)))
```

**Review by jcrist — APPROVED** [review]
Overall looks good to me! One request for a docstring example, and a few opinion-based stylistic nits (feel free to ignore).

**Comment by jcrist** [comment]
Gah, everyone reviewed this at the same time :grimacing:.

File: ibis/backends/base/sql/alchemy/registry.py (hunk 6 above)
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py:518]: I prefer direct `op.` attribute references too, it shouldn't have a performance impact due to the slotted access, but helps with code readability.

File: ibis/expr/types/relations.py (hunk 3 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:2027]: Done!

File: ibis/backends/polars/compiler.py (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/polars/compiler.py:1166]: SGTM!

File: ibis/backends/tests/test_aggregation.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/tests/test_aggregation.py]: Removed!

**Pull request merged**